### PR TITLE
New sub-lib: Trakeva_of_uri

### PR DESCRIPTION
**Pull request by smondet**
Intense first class modules and build system hacking …

[Review on Reviewable](https://reviewable.io/reviews/smondet/trakeva/18)

### CODE REVIEW

**Comment by smondet** [comment]
Running Ketrew over Postgresql here with https://github.com/hammerlab/ketrew/commit/2256f19e654450c50fb7165953d0b1a8520e38fb

File: tools/codegen.sh [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+cat <<EOBLOB >> $lib_ml
+type t = {
+  implementation: (module Trakeva.KEY_VALUE_STORE with type t = t);
+}
+let load s =
+  let uri = Uri.of_string s in
+  let backend =
+     match Uri.scheme uri with
+     | Some "postgresql" -> (module Postgresql : Trakeva.KEY_VALUE_STORE)
+     | Some "sqlite" | None -> (module Sqlite : Trakeva.KEY_VALUE_STORE)
```
Comment (rleonid) [inline tools/codegen.sh]: Why default to `Sqlite` if can't figure out uri? Shouldn't `trakeva_of_uri` then fail instead?

**Comment by rleonid** [comment]
Modulo the points I've made above it looks good. It is a real shame that you have to jump through such hoops to get it to build. I wonder if it would be possible to simplify the dependencies by asking ahead of time of the libraries are present via `ocamlfind query sqlite3` (or similar)

File: tools/codegen.sh (hunk 1 above)
Comment (smondet) [inline tools/codegen.sh]: `None` means that the URI does not have a scheme, so for backwards compatibility we assume it's a path that Sqlite can handle.

**Comment by smondet** [comment]
I'm happy to let `opam` decide wether to build the optional libraries or not.

The point of `trakeva_of_uri` is to be able to use trakeva without forcing dependencies on one or another. in the branch https://github.com/hammerlab/ketrew/tree/pluggable-db you can now build Ketrew with sqlite or postgresql or both (or none, but then it just fails because for Ketrew the DB is not optional).

File: tools/codegen.sh (hunk 1 above)
Comment (rleonid) [inline tools/codegen.sh]: Gotcha. I'd add "Defaults to interpreting path as sqlite handles if no URI specified."

Comment (smondet) [inline tools/codegen.sh]: It's there: https://github.com/smondet/trakeva/pull/18/files#diff-a52c57dea9110ea64a0ce5ff945c6516R101

Comment (rleonid) [inline tools/codegen.sh]: :+1:

**Pull request merged**